### DISCUSSION TITLE
TestOpen failing after creating 32770 sockets

### DISCUSSION THREAD

**Discussion by dormail** [ANLAB-KAIST/KENSv3#56] [Project Q&A 2. Transport Layer (TCP)]
Hi, when running kens-part1, I run into the problem of the program crashing after 32770 sockets. When changing the beginning of the test to something like this
```
    const static int test_size = 128;
    const static int test_repeat = 128;
```
It has no problem completing the test. I first though the problem might lay in my implementation of some file descriptor stuff but it is all deep inside the KENS source code which we are not supposed to change right (like functions we can not overwrite, files we are not supposed to hand in) 
Or did I simply not understand the assignment correctly?

**Comment by Pusnow** [comment]
https://github.com/ANLAB-KAIST/KENSv3/blob/0226d32cc49a1a2df8c2a3d613ba8e7509502549/app/kens/testopen.cpp#L36-L55


You might have to implement `close()` system call (line 53).
Also, you should use  file descriptor helper functions: https://github.com/ANLAB-KAIST/KENSv3/wiki/Tip:-Managing-File-Descriptors .

**Comment by dormail** [comment]
Yes, I implemented a close() call which also utilized the File Descriptor methods from KENS, the problem still persists. What I found is, that evonthough the sockets get closed, the counter for the file desriptor keeps rizing (thats how I got the exact number) so I am not sure if the deletion is working as intended.

**Reply by Pusnow** [reply]
This is how to use the file descriptor method:

```c++
  int fd[6];
  fd[0] = createFileDescriptor(pid); // returns 3
  fd[1] = createFileDescriptor(pid); // returns 4
  fd[2] = createFileDescriptor(pid); // returns 5
  fd[3] = createFileDescriptor(pid); // returns 6
  fd[4] = createFileDescriptor(pid); // returns 7
  printf("fd: %d %d %d %d %d\n", fd[0], fd[1], fd[2], fd[3], fd[4]);

  removeFileDescriptor(pid, fd[2]); // remove fd(5)

  fd[5] = createFileDescriptor(pid); // returns 5
  printf("fd[5]: %d\n", fd[5]);


  removeFileDescriptor(pid, fd[0]); // remove fd(3)
  removeFileDescriptor(pid, fd[1]); // remove fd(4)
  removeFileDescriptor(pid, fd[3]); // remove fd(6)
  removeFileDescriptor(pid, fd[4]); // remove fd(7)
  removeFileDescriptor(pid, fd[5]); // remove fd(5)


  fd[0] = createFileDescriptor(pid); // returns 3
  fd[1] = createFileDescriptor(pid); // returns 4
  fd[2] = createFileDescriptor(pid); // returns 5
  fd[3] = createFileDescriptor(pid); // returns 6
  fd[4] = createFileDescriptor(pid); // returns 7
  printf("fd: %d %d %d %d %d\n", fd[0], fd[1], fd[2], fd[3], fd[4]);
```

Output:
```
fd: 3 4 5 6 7
fd[5]: 5
fd: 3 4 5 6 7
```

As you can see, if you use the methods properly, file descriptor would not keep rising.

**Reply by dormail** [reply]
Hi, thank you I now understood what I did wrong, thank you for your help!
